manager/state/raft/transport/peer.go

```diff
@@ -0,0 +1,153 @@
+package transport
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc"
+
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/raft/membership"
+	"github.com/pkg/errors"
+)
+
+type peer struct {
+	id uint64
+
+	tr *Transport
+
+	msgc chan raftpb.Message
+
+	stopped chan struct{}
+	done    chan struct{}
+
+	mu   sync.Mutex
+	cc   *grpc.ClientConn
+	addr string
+
+	active       bool
+	becameActive time.Time
+}
+
+func newPeer(id uint64, addr string, tr *Transport) (*peer, error) {
+	cc, err := tr.dial(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create conn for %x with addr %s", id, addr)
+	}
+	return &peer{
+		id:           id,
+		addr:         addr,
+		cc:           cc,
+		tr:           tr,
+		msgc:         make(chan raftpb.Message, 4096),
+		stopped:      make(chan struct{}),
+		done:         make(chan struct{}),
+		active:       true,
+		becameActive: time.Now(),
+	}, nil
+}
+
+func (p *peer) send(ctx context.Context, m raftpb.Message) (err error) {
+	defer func() {
+		if err != nil {
+			p.tr.config.TransportRaft.ReportUnreachable(m.To)
+			p.mu.Lock()
+			p.active = false
+			p.mu.Unlock()
+		}
+	}()
+	select {
+	case p.msgc <- m:
+	case <-p.stopped:
+		return errors.New("peer stopped")
+	case <-ctx.Done():
+		return ctx.Err()
```

Comment:
`<-p.stopped` and `<-ctx.Done` should be checked in a different `select` from `p.msgc <- m`. Otherwise the `p.msgc <-m` branch can be randomly chosen.